File: JS_OC_Swift/JS_OC/JS_MessageHandler/JS_MessageHandler/ViewController.m
//
//  ViewController.m
//  JavaScriptCoreDemo
//
//  Created by xinyu on 2018/3/26.
//  Copyright © 2018年 MaChat. All rights reserved.
//

#import "ViewController.h"
#import <WebKit/WebKit.h>

@interface ViewController ()<WKScriptMessageHandler, WKNavigationDelegate, WKUIDelegate>
@property (nonatomic, strong) WKWebView *webView;
@property (nonatomic, strong) UIProgressView  *progressView;
@end

@implementation ViewController

- (void)dealloc {
    [self.webView removeObserver:self forKeyPath:@"estimatedProgress"];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    WKWebViewConfiguration *config = [[WKWebViewConfiguration alloc] init];
    // 添加自适应屏幕宽度js调用的方法
    NSString *jSString = @"var meta = document.createElement('meta'); meta.setAttribute('name', 'viewport'); meta.setAttribute('content', 'width=device-width'); document.getElementsByTagName('head')[0].appendChild(meta);";
    WKUserScript *wkUserScript = [[WKUserScript alloc] initWithSource:jSString injectionTime:WKUserScriptInjectionTimeAtDocumentEnd forMainFrameOnly:YES];
    WKUserContentController *wkUController = [[WKUserContentController alloc] init];
    [wkUController addUserScript:wkUserScript];
    
    config.userContentController = wkUController;
    
    self.webView = [[WKWebView alloc] initWithFrame:CGRectZero configuration:config];
    self.webView.frame = CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height);
    [self.view addSubview:self.webView];
    
    self.webView.UIDelegate = self;
    self.webView.navigationDelegate = self;
    
    NSString *urlStr = [[NSBundle mainBundle] pathForResource:@"index.html" ofType:nil];
    NSURL *fileURL = [NSURL fileURLWithPath:urlStr];
    [self.webView loadFileURL:fileURL allowingReadAccessToURL:fileURL];
    //进度条
    [self initProgressView];
    [self.webView addObserver:self forKeyPath:@"estimatedProgress" options:NSKeyValueObservingOptionNew context:nil];
    
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self.webView.configuration.userContentController addScriptMessageHandler:self name:@"showMessageAction"];
    [self.webView.configuration.userContentController addScriptMessageHandler:self name:@"showTitleAndMessageAction"];
    [self.webView.configuration.userContentController addScriptMessageHandler:self name:@"doSomethingThenCallBackAction"];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    [self.webView.configuration.userContentController removeScriptMessageHandlerForName:@"showMessageAction"];
    [self.webView.configuration.userContentController removeScriptMessageHandlerForName:@"showTitleAndMessageAction"];
    [self.webView.configuration.userContentController removeScriptMessageHandlerForName:@"doSomethingThenCallBackAction"];
}

- (void)initProgressView {
    CGFloat kScreenWidth = [[UIScreen mainScreen] bounds].size.width;
    UIProgressView *progressView = [[UIProgressView alloc] initWithFrame:CGRectMake(0, 0, kScreenWidth, 2)];
    progressView.tintColor = [UIColor redColor];
    progressView.trackTintColor = [UIColor lightGrayColor];
    [self.view addSubview:progressView];
    self.progressView = progressView;
}

#pragma mark - KVO
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    // 计算wkWebView进度条
    if (object == self.webView && [keyPath isEqualToString:@"estimatedProgress"]) {
        CGFloat newprogress = [[change objectForKey:NSKeyValueChangeNewKey] doubleValue];
        if (newprogress == 1) {
            [self.progressView setProgress:1.0 animated:YES];
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.7 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                self.progressView.hidden = YES;
                [self.progressView setProgress:0 animated:NO];
            });
            
        } else {
            self.progressView.hidden = NO;
            [self.progressView setProgress:newprogress animated:YES];
        }
    }
}

#pragma mark - WKScriptMessageHandler
- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message {
    
    __weak typeof(self) weakSelf = self;
    // JS 调用OC 方法  传递一个参数
    if ([message.name isEqualToString:@"showMessageAction"]) {
        UIAlertController *alertCtr = [UIAlertController alertControllerWithTitle:@"提示" message:message.body preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil];
        [alertCtr addAction:cancel];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf presentViewController:alertCtr animated:YES completion:nil];
        });
    }
    // JS 调用OC 方法  传递多个参数
    if ([message.name isEqualToString:@"showTitleAndMessageAction"]) {
        NSDictionary *dict = message.body;
        UIAlertController *alertCtr = [UIAlertController alertControllerWithTitle:dict[@"title"] message:dict[@"content"] preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil];
        [alertCtr addAction:cancel];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf presentViewController:alertCtr animated:YES completion:nil];
        });
    }
    
    if ([message.name isEqualToString:@"doSomethingThenCallBackAction"]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            UIAlertController *alertCtr = [UIAlertController alertControllerWithTitle:@"提示" message:nil preferredStyle:UIAlertControllerStyleAlert];
            [alertCtr addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
                textField.placeholder = @"请输入传入的数据!";
                [textField addTarget:weakSelf action:@selector(textFieldChange:) forControlEvents:UIControlEventEditingChanged];
            }];
            UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:nil];
            [alertCtr addAction:cancel];
            UIAlertAction *ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                UITextField *textField = [alertCtr.textFields firstObject];
                // OC 调用JS 方法  并传递参数
                NSString *jsStr = [NSString stringWithFormat:@"callback('%@')",textField.text];
                [self.webView evaluateJavaScript:jsStr completionHandler:^(id _Nullable result, NSError * _Nullable error) {
                    NSLog(@"%@----%@",result, error);
                }];
            }];
            [alertCtr addAction:ok];
            ok.enabled = NO;
            
            [weakSelf presentViewController:alertCtr animated:YES completion:nil];
        });
    }
}

- (void)textFieldChange:(UITextField *)sender {
    UIAlertController *alertCtr = (UIAlertController *)self.presentedViewController;
    UIAlertAction *okAction = [alertCtr.actions lastObject];
    if (sender.text.length > 0) {
        okAction.enabled = YES;
    } else {
        okAction.enabled = NO;
    }
}
@end

File: JS_OC_Swift/JS_Swift/JS_MessageHandler/JS_MessageHandler/ViewController.swift
//
//  ViewController.swift
//  JS_MessageHandler
//
//  Created by xinyu on 2018/3/28.
//  Copyright © 2018年 MaChat. All rights reserved.
//

import UIKit
import WebKit

class ViewController: UIViewController,WKUIDelegate,WKNavigationDelegate,WKScriptMessageHandler {

   weak var webView :WKWebView!
   weak var progressView :UIProgressView!
    
    deinit {
        webView.removeObserver(self, forKeyPath: "estimatedProgress")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupWebView()
        //显示加载进度条
        setupProgressView()
        webView.addObserver(self, forKeyPath: "estimatedProgress", options: .new, context: nil)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        webView.configuration.userContentController.add(self, name: "showMessageAction")
        webView.configuration.userContentController.add(self, name: "showTitleAndMessageAction")
        webView.configuration.userContentController.add(self, name: "doSomethingThenCallBackAction")
       
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "showMessageAction")
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "showTitleAndMessageAction")
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "doSomethingThenCallBackAction")
    }
    
    func setupWebView () {
        let config = WKWebViewConfiguration.init()
        let wkUController = WKUserContentController.init()
        config.userContentController = wkUController
        webView = WKWebView.init(frame: .zero, configuration: config)
        webView.frame = CGRect(x: 0, y: 0, width: view.frame.size.width, height: view.frame.size.height)
        view.addSubview(webView)
        webView.uiDelegate = self
        webView.navigationDelegate = self
        
        let url = URL(fileURLWithPath: Bundle.main.path(forResource: "index.html", ofType: nil)!)
        webView.loadFileURL(url, allowingReadAccessTo: url)
        
    }
   
    func setupProgressView () {
        let sWidth = UIScreen.main.bounds.width
        let pre = UIProgressView(frame: CGRect(x: 0, y: 0, width: sWidth, height: 2))
        progressView = pre
        progressView.tintColor = UIColor.blue
        progressView.trackTintColor = UIColor.lightGray
        view.addSubview(progressView)
    }
    
    //MARK: - KVO
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        
        if webView.isEqual(object) && keyPath == "estimatedProgress" {
            let newprogress = change?[NSKeyValueChangeKey.newKey] as! Float
          
            if newprogress == Float("1") {
                progressView.setProgress(1.0, animated: true)
                DispatchQueue.main.asyncAfter(deadline:.now() + 0.7, execute: {
                    DispatchQueue.main.async {
                        self.progressView.isHidden = true
                        self.progressView.setProgress(0, animated: true)
                    }
                })
            } else {
                progressView.isHidden = false
                progressView.setProgress(newprogress, animated: true)
            }
            
        }
    }
    //MARK: - WKScriptMessageHandler
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
         // JS 调用OC 方法  传递一个参数
        if message.name == "showMessageAction" {
            showAlertControlelr(title: "提示", message: message.body as! String)
        }
        // JS 调用OC 方法  传递多个参数
        if message.name == "showTitleAndMessageAction" {
            let dict : NSDictionary = message.body as! NSDictionary
            let title = dict.value(forKey: "title")
            let content = dict.value(forKey: "content")
            showAlertControlelr(title: title as! String, message: content as! String)
        }
        
        if  message.name == "doSomethingThenCallBackAction" {
            let alert = UIAlertController(title: "提示", message: nil, preferredStyle: .alert)
            alert.addTextField { (textField) in
                textField.placeholder = "请输入传送的数据"
                textField.addTarget(self, action: #selector(self.textFieldChange(textField:)), for:.editingChanged)
            }
            let cancel = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
            alert.addAction(cancel)
            let okAction = UIAlertAction(title: "OK", style: .default) { (action) in
                let textField = alert.textFields?.first
                // OC 调用JS 方法  并传递参数
                let jsStr = String(format:"callback('%@')",(textField?.text)!)
                self.webView.evaluateJavaScript(jsStr as String, completionHandler: { (result:Any?, error:Error?) in
                    print("error:",error as Any)
                })
            }
            alert.addAction(okAction)
            okAction.isEnabled = false
            self.present(alert, animated: true, completion: nil)
        }
    }
    //MARK: -
    @objc func textFieldChange(textField:UITextField) {
        let alert = presentedViewController as! UIAlertController
        let okAction = alert.actions.last
        let count = textField.text!.count
        if (count > 0) {
            okAction?.isEnabled = true
        } else {
            okAction?.isEnabled = false
        }
    }
    
    func showAlertControlelr(title:String , message:String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let cancel = UIAlertAction(title: "OK", style: .cancel, handler: nil)
        alert.addAction(cancel)
        present(alert, animated: true, completion: nil)
    }
}

